lib/hpke/src/certificate.cpp

```diff
@@ -25,14 +33,120 @@ struct Certificate::ParsedCertificate
+        throw openssl_error();
+      }
+      skid.assign(*kid);
+    }
+    return skid;
+  }
+
+  // Parse Authority Key Identifier
+  static std::string parse_akid(X509* cert)
+  {
+    int loc = X509_get_ext_by_NID(cert, NID_authority_key_identifier, -1);
+    auto* ext = X509_get_ext(cert, loc);
+    std::string akid;
+    if (ext != nullptr) {
+      auto ext_value = make_typed_unique(
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-type-reinterpret-cast)
+        reinterpret_cast<AUTHORITY_KEYID*>(X509V3_EXT_d2i(ext)));
+      auto* key_id = ext_value->keyid;
+      auto kid =
+        std::make_unique<char*>(hex_to_string(key_id->data, key_id->length));
+      if (kid == nullptr) {
+        throw openssl_error();
+      }
+      akid.assign(*kid);
+    }
+    return akid;
+  }
+
+  static ParsedSANInfo parse_san(X509* cert)
+  {
+    ParsedSANInfo san_info;
+    int san_names_nb = -1;
+    STACK_OF(GENERAL_NAME)* san_names = nullptr;
+
+    // Try to extract the names within the SAN extension from the certificate
+    san_names = static_cast<STACK_OF(GENERAL_NAME)*>(
+      X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr));
+    if (san_names == nullptr) {
+      return san_info;
+    }
+    san_names_nb = sk_GENERAL_NAME_num(san_names);
+
+    // Check each name within the extension
+    for (int i = 0; i < san_names_nb; i++) {
+      const GENERAL_NAME* current_name = sk_GENERAL_NAME_value(san_names, i);
+
+      if (current_name->type == GEN_DNS) {
+        // NOLINTNEXTLINE (cppcoreguidelines-pro-type-cstyle-cast)
+        const char* dns_name = (const char*)(ASN1_STRING_get0_data(
+          current_name->d.dNSName)); // NOLINT
+
+        // Make sure there isn't an embedded NUL character in the DNS name
+        if (ASN1_STRING_length(current_name->d.dNSName) != // NOLINT
+            static_cast<int>(strlen(dns_name))) {
+          throw std::runtime_error("Malformed certificate");
+        }
+        san_info.domains.emplace_back(std::string(dns_name));
+      } else if (current_name->type == GEN_EMAIL) {
+        const char* email =
+          // NOLINTNEXTLINE(cppcoreguidelines-pro-type-cstyle-cast)
+          (const char*)ASN1_STRING_get0_data(current_name->d.dNSName); // NOLINT
+
+        // Make sure there isn't an embedded NUL character in the DNS name
+        if (ASN1_STRING_length(current_name->d.dNSName) != // NOLINT
+            static_cast<int>(strlen(email))) {
+          throw std::runtime_error("Malformed certificate");
+        }
+        san_info.email_addresses.emplace_back(email);
+      }
+    }
+
+    // Clean up
+    sk_GENERAL_NAME_pop_free(san_names, GENERAL_NAME_free);
+    return san_info;
+  }
+
   explicit ParsedCertificate(X509* x509_in)
     : x509(x509_in, typed_delete<X509>)
     , sig_id(signature_algorithm(x509.get()))
+    , issuer(X509_NAME_oneline(X509_get_issuer_name(x509_in), nullptr, 0))
+    , subject(X509_NAME_oneline(X509_get_subject_name(x509_in), nullptr, 0))
```

Comment:
This is broken.  It makes your issuer/subject matching depend on OpenSSL's stringification of distinguished names.  Instead, you want to use [the "canonical encoding" that is already provided](https://github.com/openssl/openssl/blob/a829b735b645516041b55746e013692babd8cd31/include/crypto/x509.h#L38).

Reply:
might need some help on the right accessor on this